pyrsistent/_pmap.py

```diff
@@ -125,7 +125,48 @@ def __len__(self):
     def __repr__(self):
         return 'pmap({0})'.format(str(dict(self)))
 
-    __eq__ = Mapping.__eq__
+    def __eq__(self, other):
+        if self is other:
+            return True
+        if not isinstance(other, Mapping):
+            return NotImplemented
+        if len(self) != len(other):
+            return False
+        if isinstance(other, PMap):
+            if (hasattr(self, '_cached_hash') and hasattr(other, '_cached_hash')
+                    and self._cached_hash != other._cached_hash):
+                return False
+            if self._buckets == other._buckets:
+                return True
+            self_as_set = other_as_set = self_hash = other_hash = None
+            if hasattr(self, '_cached_hash'):
```

Comment:
From here and below I feel like this method is starting to get a bit out of hand. What used to be a one-liner is now 40 lines of if-elses... :-)

The code above should sort out a lot of cases where the maps are different. I imagine the below adds very little in terms of performance and actually makes a fairly common case (that the maps are equal) slower while adding a lot of code.

I would probably remove everything down to line 166.